src/ngraph/pass/algebraic_simplification.cpp

```diff
@@ -487,6 +488,66 @@ static bool simplify_log(shared_ptr<Node> n)
     return false;
 }
 
+// optimizes `gather->shapeof` into `shapeof->gather` for 0D indices
+// other cases into Concat of shapeof/gather(data) + shapeof(indices)
+static bool simplify_gather_shapeof(shared_ptr<Node> node)
+{
+    auto shapeof = as_type_ptr<op::v0::ShapeOf>(node);
+    auto gather = as_type_ptr<op::v1::Gather>(shapeof->input_value(0).get_node_shared_ptr());
+    if (!gather)
+    {
+        return false;
+    }
+    auto gather_in_rank = gather->get_input_partial_shape(0).rank();
+    auto indices_rank = gather->get_input_partial_shape(1).rank();
+    auto axis = gather->get_axis();
+    if (gather_in_rank.is_dynamic() || indices_rank.is_dynamic() ||
+        axis == op::v1::Gather::AXIS_NOT_SET_VALUE)
+    {
+        NGRAPH_DEBUG << gather << " cannot simplify gather->shapeof";
+        return false;
+    }
+
+    auto zero_axis = op::Constant::create<int64_t>(element::i64, Shape{}, {0});
+    shared_ptr<Node> replace_node;
+    auto new_shapeof = make_shared<op::v0::ShapeOf>(gather->input_value(0).get_node_shared_ptr());
```

Comment:
As we are going to use this transformation in our common optimizations pass we need to be sure that all operations produced by this transformation belongs to opset1 or higher. 
So using v0/v1/vXXX namespace complicates transformation review and I can't 100% say that all this ops are from opset1 or higher. That's why opsetXXX namespace was introduced. In case of new transformation we use opset3 now.

Reply:
technically opset3::xx will resolve to same underlying op, anyways i made the changes to use opset3 namespace.